### ISSUE TITLE
[java] Enhance UseStandardCharsets to flag some constructors of IO-related classes 

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3212]
This is extracted from #3190, fyi @aaime 

**Is your feature request related to a problem? Please describe.** The rule UseStandardCharsets (#3193) flags explicit use of `Charset::forName`. 

However, there is a number of Java API where an overloaded method takes the Charset defition as a string, e.g. the following should also be flagged and reported:

```java
       try (OutputStreamWriter osw = new OutputStreamWriter(out, "UTF-8")) { }
```

**Describe the solution you'd like** In pmd 7 this is easy to solve. We can use invocation matchers in XPath: `ConstructorCall[pmd-java:matchesSig('java.io.OutputStreamWriter#new(_,java.lang.String)')]`. Perhaps writing the rule in java would be better though, if there are many ctors to match.

**Describe alternatives you've considered** Doing this on master. But this would expose us to false positives I think, as these constructors have many overloads and I don't think the overload resolver is very reliable. Since this rule is part of `quickstart.xml`, I don't want to make concessions about precision.

**Comment by dan-s1** [comment]
Is this issue still being worked? I have some additional examples I would expect to be flagged:

1. Base64.getDecoder().decode("someString"), "UTF-8")
2. URLEncoder.encode("someString", "UTF-8")
3. "someString".getBytes("UTF-8")
4. new String(someByteArray, "UTF-8")
5. intanceOfByteArrayOutputStream.toString("UTF-8")

These are actually flagged within Intellij and I would expect when using the Maven PMD plugin these should also be flagged.

**Comment by UncleOwen** [comment]
I've created a list of all public methods/constructors I could find in java.base:
* String(byte bytes[], int offset, int length, String charsetName)
* String(byte bytes[], String charsetName)
* String.getBytes(String charsetName)
* Scanner(ReadableByteChannel source, String charsetName)
* Scanner(Path source, String charsetName)
* Scanner(File source, String charsetName)
* Scanner(InputStream source, String charsetName)
* Properties.storeToXML(OutputStream os, String comment, String encoding)
* Formatter(OutputStream os, String csn, Locale l)
* Formatter(File file, String csn, Locale l)
* Formatter(String fileName, String csn, Locale l)
* PrintWriter(String fileName, String csn)
* PrintWriter(File file, String csn)
* PrintStream(OutputStream out, boolean autoFlush, String encoding)
* PrintStream(String fileName, Charset charset)
* PrintStream(File file, String csn)
* InputStreamReader(InputStream in, String charsetName)
* OutputStreamWriter(OutputStream out, String charsetName)
* ByteArrayOutputStream.String toString(String charsetName)
* Channels.newWriter(WritableByteChannel ch, String csName)
* Channels.newReader(ReadableByteChannel ch, String csName)
* URLDecoder.decode(String s, String enc)
* URLEncoder.encode(String s, String enc)

Each of these has an equivalent that takes a Charset instead of a String.

**Comment by adangel** [comment]
This list might be similar to the one in #2186
